lib/src/package_config_json.dart

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "dart:convert";
+import "dart:io";
+import "dart:typed_data";
+
+import "package:path/path.dart" as path;
+
+import "package_config_impl.dart";
+import "packages_file.dart" as packages_file;
+import "util.dart";
+
+const String _configVersionKey = "configVersion";
+const String _packagesKey = "packages";
+const List<String> _topNames = [_configVersionKey, _packagesKey];
+const String _nameKey = "name";
+const String _rootUriKey = "rootUri";
+const String _packageUriKey = "packageUri";
+const String _languageVersionKey = "languageVersion";
+const List<String> _packageNames = [
+  _nameKey,
+  _rootUriKey,
+  _packageUriKey,
+  _languageVersionKey
+];
+
+const String _generatedKey = "generated";
+const String _generatorKey = "generator";
+const String _generatorVersionKey = "generatorVersion";
+
+/// Reads a package configuration file.
+///
+/// Detects whether the [file] is a version one `.packages` file or
+/// a version two `package_config.json` file.
+///
+/// If the [file] is a `.packages` file, first checks whether there is an
+/// adjacent `.dart_tool/package_config.json` file, and if so,
+/// reads that instead.
+///
+/// The file must exist and be a normal file.
+Future<PackageConfig> readAnyConfigFile(File file) async {
+  var bytes = await file.readAsBytes();
+  for (int i = 0; i < bytes.length; i++) {
+    var char = bytes[i];
+    if (char == 0x20 || char == 0x09 || char == 0x0a || char == 0x0d) {
+      // Skip whitespace.
+      continue;
+    }
+    if (char == 0x7b /*{*/) {
+      // Probably JSON, definitely not .packages.
+      return parsePackageConfigBytes(bytes, file);
+    }
+    break;
+  }
+  // File is not JSON.
+  var alternateFile = File(
+      path.join(path.dirname(file.path), ".dart_tool", "package_config.json"));
+  if (alternateFile.existsSync()) {
+    return parsePackageConfigBytes(
+        await alternateFile.readAsBytes(), alternateFile);
+  }
+  return _parseDotPackagesConfig(bytes, file);
+}
+
+Future<PackageConfig> readPackageConfigJsonFile(File file) async =>
+    parsePackageConfigBytes(await file.readAsBytes(), file);
+
+Future<PackageConfig> readDotPackagesFile(File file) async =>
+    packages_file.parse(await file.readAsBytes(), Uri.file(file.path));
+
+PackageConfig parsePackageConfigBytes(Uint8List bytes, File file) {
+  // TODO(lrn): Make this simpler. Maybe parse directly from bytes.
+  return parsePackageConfigJson(
+      json.fuse(utf8).decode(bytes), Uri.file(file.path));
+}
+
+/// Creates a [PackageConfig] from a parsed JSON-like object structure.
+///
+/// The [json] argument must be a JSON object (`Map<String, dynamic>`)
```

Comment:
Can we type the argument as `Map<String, dynamic>` instead of `dynamic`?

Reply:
It's going to be called with the result of `json.decode` which may or may not be a `Map<String, dynamic>`. We might as well include the check for being a `Map<String, dynamic>` here instead of having to do it at each call-site before calling.